In [1]:
import numpy as np
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis as QDA
'''
Function to test the correct format of list of list

Return: Boolen
'''
def test_dataset(data):
    if len(data) != 150:
        return False
   
    for row in data:
        if len(row) != 5:
            return False
       
        for column in row[:-1]:
            if type(column) != np.float64:
                print("not float is a ", type(column), column)
                return False
           
        if type(row[-1]) != str:
            print("not a string")
            return False
   
    return True

data = []

'''
problem1 function read the data, and format the data to numpy array

Return: the numpy array
'''
def problem1():

    ''''load data from .data file to list of list'''
    f=open("iris.data")
    sample=f.readlines()
    for i in range(len(sample)):
        sample[i] = sample[i].strip().split(',')

    for item in sample:
        list=[]
        for i in range(len(item)):
            if i!=(len(item)-1):
                list.append(np.float64(item[i]))
            else:
                list.append(item[i])
        data.append(list)

    print("the test_dataset function returns:",test_dataset(data))
    '''convert data to numpy array'''  
    prepared_data = np.array(data)
    
    return prepared_data

'''
Function parse the data into numpy arrays of data and labels:

Return the numpy data and labels
'''   
def parse_data(data):
    training_data = []
    testing_data = []
    training_data_labels=[]
    testing_data_labels=[]
    '''separate training data and testing data into list of list forms'''
    for i in range(len(data)):
        if i in range(0,40) or i in range(50,90) or i in range(100, 140):
            training_data.append(data[i])
        else:
            testing_data.append(data[i])


    #print(training_data,"\n\n\n\n" ,testing_data)

    '''separate the labels and data in list of list forms'''
    for ii in range(len(training_data)):
        training_data_labels.append(training_data[ii][-1])
        training_data[ii]=training_data[ii][:len(training_data[ii])-1]


    for jj in range(len(testing_data)):
        testing_data_labels.append(testing_data[jj][-1])
        testing_data[jj]=testing_data[jj][:len(testing_data[jj])-1]


    '''convert everything into numpy array'''
    training_data = np.array(training_data)
    training_data_labels = np.array(training_data_labels)
    testing_data = np.array(testing_data)
    testing_data_labels = np.array(testing_data_labels)
    print("No column has been removed:")
    return training_data,training_data_labels,testing_data,testing_data_labels

In [2]:
'''
Function problem2 train the data with LDA

Print out testing and training error 

Return the LDA object
'''
def problem2():
    '''LDA test with testing data'''
    lda = LDA()
    lda_test_re = lda.fit(training_data,training_data_labels).score(testing_data,testing_data_labels)
    lda_train_re = lda.fit(training_data,training_data_labels).score(training_data,training_data_labels)
    print("LDA testing error rate is", (1-lda_test_re)*100,"%")
    print("LDA training error rate is", (1-lda_train_re)*100,"%")
    return lda

In [3]:
 '''
 Function problem3 train the data with QDA
 
 Print out testing and training error
 
 Return the QDA object
 '''
def problem3():
    '''QDA test with testing data'''
    qda = QDA()
    qda_test_re = qda.fit(training_data,training_data_labels).score(testing_data,testing_data_labels)
    print("QDA prediction is:", qda.predict(testing_data) )
    qda_train_re = qda.fit(training_data,training_data_labels).score(training_data,training_data_labels)
    print("QDA testing error rate is", (1-qda_test_re)*100,"%")
    print("QDA training error rate is", (1-qda_train_re)*100,"%\n\n")
    return qda

In [4]:
#print(problem1())
problem1()
#training_data,training_data_labels,testing_data,testing_data_labels = parse_data(data)
#problem2()
#problem3()

the test_dataset function returns: True


array([['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'],
       ['4.9', '3.0', '1.4', '0.2', 'Iris-setosa'],
       ['4.7', '3.2', '1.3', '0.2', 'Iris-setosa'],
       ['4.6', '3.1', '1.5', '0.2', 'Iris-setosa'],
       ['5.0', '3.6', '1.4', '0.2', 'Iris-setosa'],
       ['5.4', '3.9', '1.7', '0.4', 'Iris-setosa'],
       ['4.6', '3.4', '1.4', '0.3', 'Iris-setosa'],
       ['5.0', '3.4', '1.5', '0.2', 'Iris-setosa'],
       ['4.4', '2.9', '1.4', '0.2', 'Iris-setosa'],
       ['4.9', '3.1', '1.5', '0.1', 'Iris-setosa'],
       ['5.4', '3.7', '1.5', '0.2', 'Iris-setosa'],
       ['4.8', '3.4', '1.6', '0.2', 'Iris-setosa'],
       ['4.8', '3.0', '1.4', '0.1', 'Iris-setosa'],
       ['4.3', '3.0', '1.1', '0.1', 'Iris-setosa'],
       ['5.8', '4.0', '1.2', '0.2', 'Iris-setosa'],
       ['5.7', '4.4', '1.5', '0.4', 'Iris-setosa'],
       ['5.4', '3.9', '1.3', '0.4', 'Iris-setosa'],
       ['5.1', '3.5', '1.4', '0.3', 'Iris-setosa'],
       ['5.7', '3.8', '1.7', '0.3', 'Iris-setosa'],
       ['5.1

In [5]:
'''Function below used to test out problem 4'''

'''
Function that parse data to training, testing data and labels

Parameter: data --> the list of list 
            feature_to_remove ---> the number of column to remove from data sample

Returns the new training, testing data and corresponding labels
'''
def new_parse_data(data,feature_to_remove):
    new_training_data = []
    new_testing_data = []
    new_training_data_labels=[]
    new_testing_data_labels=[]
    '''separate training data and testing data into list of list forms'''
    for i in range(len(data)):
        if i in range(0,40) or i in range(50,90) or i in range(100, 140):
            new_training_data.append(data[i])
        else:
            new_testing_data.append(data[i])


    '''separate the labels and data in list of list forms'''
    for ii in range(len(new_training_data)):
        new_training_data_labels.append(new_training_data[ii][-1])
        new_training_data[ii]=new_training_data[ii][:len(new_training_data[ii])-1]

    

    for jj in range(len(new_testing_data)):
        new_testing_data_labels.append(new_testing_data[jj][-1])
        new_testing_data[jj]=new_testing_data[jj][:len(new_testing_data[jj])-1]

    '''remove certain feature from training data and testing data'''
    for kk in range(len(new_training_data)):
        new_training_data [kk].pop(feature_to_remove)
    
    for ll in range(len(new_testing_data)):
        new_testing_data[ll].pop(feature_to_remove)

    '''convert everything into numpy array'''
    new_training_data = np.array(new_training_data)
    new_training_data_labels = np.array(new_training_data_labels)
    new_testing_data = np.array(new_testing_data)
    new_testing_data_labels = np.array(new_testing_data_labels)
    print("Column number",feature_to_remove+1 ,"is removed:")
    
    return new_training_data,new_training_data_labels,new_testing_data,new_testing_data_labels

In [6]:
'''
Function train the data with LDA

Parameter: the data that has certain column removed

Print out the error rate of training and testing

Return the LDA object
'''
def new_problem2(training_data,training_data_labels,testing_data,testing_data_labels):
    '''LDA test with testing data'''
    lda = LDA()
    lda_test_re = lda.fit(training_data,training_data_labels).score(testing_data,testing_data_labels)
    lda_train_re = lda.fit(training_data,training_data_labels).score(training_data,training_data_labels)
    print("LDA testing error rate is", (1-lda_test_re)*100,"%")
    print("LDA training error rate is", (1-lda_train_re)*100,"%")
    return lda

In [7]:
'''
Function train the data with QDA 

Parameter: the data that has certain column removed

Print out the error rate of training and testing

Return QDA object
'''
def new_problem3(training_data,training_data_labels,testing_data,testing_data_labels):
    '''QDA test with testing data'''
    qda = QDA()
    qda_test_re = qda.fit(training_data,training_data_labels).score(testing_data,testing_data_labels)
    qda_train_re = qda.fit(training_data,training_data_labels).score(training_data,training_data_labels)
    print("QDA testing error rate is", (1-qda_test_re)*100,"%")
    print("QDA training error rate is", (1-qda_train_re)*100,"%\n\n")
    return qda

In [8]:
#problem1()
training_data,training_data_labels,testing_data,testing_data_labels = parse_data(data)
#problem2()
problem3()
'''
for i in range(4):
    new_training_data,new_training_data_labels,new_testing_data,new_testing_data_labels = new_parse_data(data,i)
    new_problem2(new_training_data,new_training_data_labels,new_testing_data,new_testing_data_labels)
    new_problem3(new_training_data,new_training_data_labels,new_testing_data,new_testing_data_labels)
'''

No column has been removed:
QDA prediction is: ['Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-virginica' 'Iris-virginica'
 'Iris-virginica' 'Iris-virginica' 'Iris-virginica' 'Iris-virginica'
 'Iris-virginica' 'Iris-virginica' 'Iris-virginica' 'Iris-virginica']
QDA testing error rate is 0.0 %
QDA training error rate is 1.6666666666666718 %




'\nfor i in range(4):\n    new_training_data,new_training_data_labels,new_testing_data,new_testing_data_labels = new_parse_data(data,i)\n    new_problem2(new_training_data,new_training_data_labels,new_testing_data,new_testing_data_labels)\n    new_problem3(new_training_data,new_training_data_labels,new_testing_data,new_testing_data_labels)\n'

Problem 4 answers: 

The data I get:

No column has been removed:
LDA testing error rate is 0.0 %
LDA training error rate is 2.500000000000002 %
QDA testing error rate is 0.0 %
QDA training error rate is 1.6666666666666718 %


Column number 1 is removed:
LDA testing error rate is 0.0 %
LDA training error rate is 1.6666666666666718 %
QDA testing error rate is 0.0 %
QDA training error rate is 2.500000000000002 %


Column number 2 is removed:
LDA testing error rate is 0.0 %
LDA training error rate is 2.500000000000002 %
QDA testing error rate is 0.0 %
QDA training error rate is 1.6666666666666718 %


Column number 3 is removed:
LDA testing error rate is 0.0 %
LDA training error rate is 5.833333333333335 %
QDA testing error rate is 0.0 %
QDA training error rate is 4.1666666666666625 %


Column number 4 is removed:
LDA testing error rate is 3.3333333333333326 %
LDA training error rate is 4.1666666666666625 %
QDA testing error rate is 3.3333333333333326 %
QDA training error rate is 5.833333333333335 %

Base on the data I collected, I would claim that feature in column number 2,which is the sepal width in cm,is not important, because it has the same LDA and QDA error rate as the one that none of feature is removed.